### ISSUE TITLE
Allow runtime-wide `TryState` checks

### ISSUE THREAD

**Issue by ggwpez** [paritytech/polkadot-sdk#235]
Currently the `TryState` can only be utilized on a per-pallet level by implementing the `try_state` hook.   
This limits the usage tor multi-pallet `try_state` functionality. `Executive` could be extended to accept an optional `TryState` which will then be executed by the try-runtime CLI on a runtime-wide level.

cc @kianenigma

**Comment by liamaharon** [comment]
> So we can already write and execute `TryState` checks in `Executive`? Then we can close this issue?

On further thought I don't think we can. Most runtimes use the `Executive` pallet from FRAME rather than defining their own. I think we'd want some way to pass try-state hooks into the executive pallet, like we do with migrations, so that chain-specific try-state checks can be written

**Comment by ggwpez** [comment]
>I just said that we don't need to move this to frame-executive, because we just can call AllPalletsWithSystem::try_state in the runtime

You mean we copy&paste the same try-state check into each runtime where we want to use it?  
I think we could for example have a multi-pallet check for `Staking` that takes a few generics and then checks that all the staking pallets have consistent overall state.

**Comment by kianenigma** [comment]
Sensing some confusion while reading this. 

So, looking at the issue again afresh, we need to do exactly what we do with 'COnRuntimeUpgrade': Pass a new 'CTryState = ()' to executive, and run all checks on '(AllPallets, CTryState)' instead. Sounds right?

**Comment by ntn-x2** [comment]
Any update on this issue? I just stumbled across a similar need, and the only way around it that I found is feature-gating an additional type parameter on each pallet's `Config`, to allow the runtime to be injected in there.

**Comment by ntn-x2** [comment]
So we could replace

```rust
if checks.try_state() {
    let _guard = frame_support::StorageNoopGuard::default();
    <AllPalletsWithSystem as frame_support::traits::TryState<BlockNumberFor<System>>>::try_state(
        frame_system::Pallet::<System>::block_number(),
	frame_try_runtime::TryStateSelect::All,
    )?;
}
```

with

```rust
if checks.try_state() {
    let _guard = frame_support::StorageNoopGuard::default();
    <(COnRuntimeUpgrade, AllPalletsWithSystem) as frame_support::traits::TryState<BlockNumberFor<System>>>::try_state(
        frame_system::Pallet::<System>::block_number(),
	frame_try_runtime::TryStateSelect::All,
    )?;
}
```

**Comment by liamaharon** [comment]
> So we could replace
> 
> ```rust
> if checks.try_state() {
>     let _guard = frame_support::StorageNoopGuard::default();
>     <AllPalletsWithSystem as frame_support::traits::TryState<BlockNumberFor<System>>>::try_state(
>         frame_system::Pallet::<System>::block_number(),
> 	frame_try_runtime::TryStateSelect::All,
>     )?;
> }
> ```
> 
> with
> 
> ```rust
> if checks.try_state() {
>     let _guard = frame_support::StorageNoopGuard::default();
>     <(COnRuntimeUpgrade, AllPalletsWithSystem) as frame_support::traits::TryState<BlockNumberFor<System>>>::try_state(
>         frame_system::Pallet::<System>::block_number(),
> 	frame_try_runtime::TryStateSelect::All,
>     )?;
> }
> ```
> 
> everywhere where relevant (this is taken from substrate v1.0, but it seems it still applies [here](https://github.com/paritytech/polkadot-sdk/blob/313fe0f9a277f27a4228634f0fb15a1c3fa21271/substrate/frame/executive/src/lib.rs#L380) and [here](https://github.com/paritytech/polkadot-sdk/blob/313fe0f9a277f27a4228634f0fb15a1c3fa21271/substrate/frame/executive/src/lib.rs#L459) as of today).

This sounds reasonable to me, similar API as runtime upgrades.

**Comment by ntn-x2** [comment]
> This sounds reasonable to me, similar API as runtime upgrades.

It would nevertheless be a breaking change since `COnRuntimeUpgrade` is not required to implement the `TryState` trait as well. Having a second generic with a default value of `()` would not be a breaking change, on the other hand, but I feel this approach would be less scalable in the long-term.

**Comment by liamaharon** [comment]
Yeah I wouldn't add it to the existing trait. Maybe a new one like `COnTryState`?

**Comment by kianenigma** [comment]
> Sensing some confusion while reading this. 
> 
> So, looking at the issue again afresh, we need to do exactly what we do with 'COnRuntimeUpgrade': Pass a new 'CTryState = ()' to executive, and run all checks on '(AllPallets, CTryState)' instead. Sounds right? 

have you tried this?

**Comment by kianenigma** [comment]
> If there's yet a new testing function, that would most likely mean a new generic to account for those

The nice thing about generics is that they can have defaults, so this is not a breaking change, as long as you add it to the end of the list. 

This is not pretty or elegant, but it is backwards compat, and near zero effort, I would do this. 

Longer term, I have shared some ideas around improving the development surface of executive with @gupnik or in other issues, but I am not sure if they are formalized or worth doing at this point.

**Comment by liamaharon** [comment]
I agree with @kianenigma this approach is worth trying. It shouldn't be breaking, `COnRuntimeUpgrade` type is actually optional on `Executive` so `CTryState` should be able to be as well.

**Comment by ntn-x2** [comment]
Yes, I agree this approach would not be breaking and relatively straightforward. I was just concerned it might not be the best, but I'm happy to hear there's discussion going on about improving DX on that side.

I'll try to work a PR shortly, and would be great to get this backported to at least as back as 1.6.

**Comment by ntn-x2** [comment]
The matter is less trivial than I was expecting. The `TryState` trait is implemented for tuples, but only where all the elements implement the `PalletInfoAccess` trait as well, which is clearly not required for a runtime-wide test: https://github.com/paritytech/polkadot-sdk/blob/c7cb1f25d1f8bb1a922d466e39ee935f5f027266/substrate/frame/support/src/traits/try_runtime/mod.rs#L159.

Also, I was hoping for a solution where whatever type wants to implement `TryState`, would not have to deal with the `targets: Select` argument of the trait, but rather implement something closer to the `try_state` function exposed as part of the `Hooks` trait.

I am wondering if declaring another trait that does not have the second argument, have a blanket impl of that trait for everything that implements the `TryState` trait except for `AllPallets*` types (there's no native way to exclude types from a blanket impl), and manually implement the new trait for `AllPallets*`. Otherwise, I see no other way to solve this issue. 

Cc @kianenigma since I see he was the one working on this feature last.

**Comment by liamaharon** [comment]
Do we know what does it use `PalletInfoAccess` for?

**Comment by ntn-x2** [comment]
@liamaharon yes, you can filter which pallets you want to run the tests for: https://github.com/paritytech/polkadot-sdk/blob/c7cb1f25d1f8bb1a922d466e39ee935f5f027266/substrate/frame/support/src/traits/try_runtime/mod.rs#L213.

**Comment by liamaharon** [comment]
How about we add a new method `id` and remove the `PalletInfoAccess` requirement?

Something like
```rust
pub trait TryState<BlockNumber> {
	/// Execute the state checks.
	fn try_state(_: BlockNumber, _: Select) -> Result<(), TryRuntimeError>;
        fn id() -> String
}
```

**Comment by ntn-x2** [comment]
@liamaharon I am not sure that would be an ideal design. The trait already requires the implementor to perform some `try_state` depending on the `checks` argument that it is passed. Having an `id() -> String` method on the trait would add unnecessary duplication. Rather, I would go with adding a new trait that contains the `id() -> String` function and require that in pallet executive rather. Then we can implement the trait for `PalletInfoAccess` implementing types as well. I am not sure there is a method within Substrate codebase that allows to return a specific type and that also takes `&self` as a parameter, something like a `GetWithArg<T>` which has a `fn get(&self) -> T` method.

What's your opinion on that?

**Comment by liamaharon** [comment]
Hm yes you are right. I missed that `id` would need to be manually specified, I initially thought it could be done in the pallet macros.

Your approach sounds good to try, let's see how it looks

**Comment by ntn-x2** [comment]
Ok I have to pause work on this again right now, but I found that returning a String ID does not work either, as it would not be implementable by the `AllPalletsWithSystem` and `AllPalletsWithoutSystem` types. I found an alternative way, which is for the tuple to simply implement an `Eq<&[u8]>` logic, and the specific `try_state` is executed if any of the provided names matches what was provided as argument.

**Comment by kianenigma** [comment]
> How about we add a new method `id` and remove the `PalletInfoAccess` requirement?
> 
> Something like
> 
> ```rust
> pub trait TryState<BlockNumber> {
> 	/// Execute the state checks.
> 	fn try_state(_: BlockNumber, _: Select) -> Result<(), TryRuntimeError>;
>         fn id() -> String
> }
> ```
> 
> When pallet implements it, it can use its name. This might be a bit tricky, let me know if u get stuck.

Maybe it is not a breaking change if we provide an auto-implementation of `fn id()` when this is being generated from within a pallet? also, if this is used just for logging, providing something like this is reasonable to me: 

> ```rust
> pub trait TryState<BlockNumber> {
> 	/// Execute the state checks.
> 	fn try_state(_: BlockNumber, _: Select) -> Result<(), TryRuntimeError>;
>         fn id() -> String { "PLEASE-PROVIDE-ID" }
> }
> ```

**Comment by ntn-x2** [comment]
@kianenigma @liamaharon I opened a PR that fixes this: https://github.com/paritytech/polkadot-sdk/pull/4631. I did not find a way to make it work by adding an `id() -> String` function to the `TryState` trait, which must be implemented for tuples, but I found a nice solution that is backwards compatible, and allows developer to hook their `try-state` logic without having to implement any logic regarding the `Select` parameter of the `try_state` function. Eager to get your feedback on it.
